### DISCUSSION TITLE
versioning python

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2028] [Writing Formulae/Casks]
i'm having a difficult time making `brew audit --formula ./Formula/python3.9.rb` happy.

In the freecad/homebrew-freecad tap we have a tap based version of python 3.9 that uses the tap python instead of homebrew-core _python@3.9_

me personally, i haven't run into many issues using both the freecad based python and the homebrew-core based python.

if i ever run into an issue, i usually check to see which version is linked my cellar and then link such python to fix/resolve the issue that arises, which is rather seldom at best.

so all that said, i'm working on building bottles using the homebrew provided github actions, however it seems that using a versioned python is causing an _audit_ error thus preventing any bottles from being built that rely on my tap specific version of python ie.

```
cd /usr/local/Homebrew/Library/Taps/freecad/homebrew-freecad
brew audit --formula ./Formula/python3.9.rb

 * Python formulae that are keg-only should not create `pip3` and `wheel3` symlinks.
```

i read the [versioning doc][1], and can't seem to avoid having the above audit error arise. doing a quick _grep_ in the homebrew _brew_ cmd repo leads me to this block,

```rb
# formula_cellar_checks.rb

  def check_python_symlinks(name, keg_only)
    return unless keg_only
    return unless name.start_with? "python"

    return if %w[pip3 wheel3].none? do |l|
      link = HOMEBREW_PREFIX/"bin"/l
      link.exist? && File.realpath(link).start_with?(HOMEBREW_CELLAR/name)
    end

    "Python formulae that are keg-only should not create `pip3` and `wheel3` symlinks."
  end
```

**how can i get my versioned formula ie. python3.9 to pass this check?**

[1]: <https://docs.brew.sh/Versions>

**Comment by carlocab** [comment]
Make it keg-only, I guess? Why do you need your own version of Python 3.9?

**Reply by ipatch** [reply]
it's just what i'm dealing with. for whatever reason that audit check was going off when the formula had the name of `python3.9` but when renamed the formula to `python@3.9` things having been moving along better. 👍

**Reply by ipatch** [reply]
> Make it keg-only, I guess? Why do you need your own version of Python 3.9?

previous maintainers of the homebrew-freecad tap decided to version a lot of the formula contained within the tap because of the frequent updates to certain formula. we were running into more issues than not when certain formula were updating ie. boost. so it was decided to version out the formula in the tap, so those building freecad from source would not be hindered by running `brew upgrade` and have breaking changes roll in when certain packages were updated.

a bigger discussion about the topic can be found below,

https://github.com/Homebrew/discussions/discussions/342

https://github.com/FreeCAD/homebrew-freecad/pull/158#issuecomment-773637997

**Comment by ipatch** [comment]
i noticed that forks of `homebrew-core` have a `audit_exceptions` dir and a `style_exceptions` dir in the root dir of the tap.

is it not possible to add a `audit exception` for the _homebrew-freecad_ version of python?

and if not? how could i modify the formula to create the `wheel` or `pip` symlinks so the _audit_ can pass without any errors or warnings?
